### DISCUSSION TITLE
ipsw outputs mangled symbol names on Monterey

### DISCUSSION THREAD

**Discussion by rgoldberg** [blacktop/ipsw#563] [Q&A] [answered]
The following command line outputs mangled symbol names using `ipsw` 3.1.544 installed via `brew install ipsw` on current macOS Monterey 12.7.6 on an Intel Mac:

```
ipsw class-dump --arch x86_64 --demangle --verbose --re --refs --headers -o ~/directory /System/Library/dyld/dyld_shared_cache_x86_64 CommerceKit
```

I've tried it using `dyld_shared_cache_x86_64h` (`h` appended) instead of `dyld_shared_cache_x86_64`.

I've tried removing all of the options other than `--headers -o ~/directory`.

I've tried other private frameworks.

The results seem deterministic since rerunning the same command line produces the same output. Using different options / values for the command, however, produces different symbol names, but all the combinations I've tried still produce mangled symbol names.

How can I use `ipsw` to output correct symbol names?

I also tried:

```
ipsw dyld macho /System/Library/dyld/dyld_shared_cache_x86_64 CommerceKit --obj
```

I tried with `x86_64` & `arm64e`, too. All three output mangled symbol names.

Here's one of the mangled files (`CKBag.h`) from one of my various attempts:

```Objective-C
//
//   Generated by https://github.com/blacktop/ipsw (Version: 3.1.544, BuildCommit: Homebrew)
//
//    - LC_BUILD_VERSION:  Platform: macOS, MinOS: 12.7, SDK: 12.7, Tool: ld (760)
//    - LC_SOURCE_VERSION: 712.6.1.0.0
//
#ifndef CKBag_h
#define CKBag_h
@import Foundation;

#include "CKStoreClient.h"

@interface CKBag : NSObject

@property (readonly, nonatomic) CKStoreClient *storeClient;

/* class methods */
+ (id)addingTableEntry;
+ (id)iriSettingsIntents19SetFocusModeHandler;
+ (id)xDotCount:;
+ (id)uldTreatAllSmallStrokesAsPunctuation;
+ (id)ation;

/* instance methods */
- (signed char)eRepositoryInterface;
- (void)st;
- (id)nt;
- (id)addingTableEntry;
- (id)iriSettingsIntents19SetFocusModeHandler;
- (id)y",R,C,N,V_otherDateComponents;
- (id)enQueueEmpty;
- (id)ultsByLocale;
- (signed char)zeCandidatesForRomanPunctuation:(id)punctuation;
- (signed char)rRomanPunctuation:(id)punctuation punctuationStrokeCount:(id)count;
- (signed char)nStrokeCount:(id)count;
- (id)argedConvexHull:withMargin:;
- (id)l:(id /* block */)l;
- (void)ithOrigin:(id)origin;
- (void)nceFromPoint:toRectangle:;

@end

#endif /* CKBag_h */
```

**Comment by blacktop** [comment]
when you say 'mangled' do you mean C++/Swift mangled or just 'incorrect'

**Reply by rgoldberg** [reply]
I just meant that some of the symbol names are obviously wrong.

You can see that many of them are truncated mid-word at the start. Also, changing the DSC gives different symbol names.

If you try variations of the command lines I sent, you'll probably see the same issue, unless thing work better on different versions of macOS or on different hardware (like on Apple Silicon).

I said mangled because I saw '--demangle`.  Maybe mangled is a different issue.

**Reply by rgoldberg** [reply]
@blacktop Have you found any way to generate headers that deterministically output correct symbol names (i.e. that never truncate the start of, or otherwise mess up, symbol names)?

**Reply by blacktop** [reply]
try on more recent version of macOS

**Reply by rgoldberg** [reply]
@blacktop I don't have access to a more recent macOS version now.

I will get an M4, though, after they are released & before macOS 12 goes EOL.

Do you know for sure that it will work on newer macOS versions?

If not, do you mind trying to run one of the command lines, then letting me know if you also experience the same problem?

If new masOS versions also have the same problem, then it seems like a bug worth fixing.

Thanks.

**Reply by blacktop** [reply]
```objc
//
//   Generated by https://github.com/blacktop/ipsw (Version: , BuildCommit: )
//
//    - LC_BUILD_VERSION:  Platform: macOS, MinOS: 15.1, SDK: 15.1, Tool: ld (1115.7.3)
//    - LC_SOURCE_VERSION: 715.1.1.0.0
//
#ifndef CKBag_h
#define CKBag_h
@import Foundation;

#include "CKStoreClient.h"

@interface CKBag : NSObject

@property (readonly, nonatomic) CKStoreClient *storeClient;

/* class methods */
+ (id)valueForKey:(id)key;
+ (id)urlForKey:(id)key;
+ (id)bagWithType:(unsigned long long)type;
+ (id)productionBag;
+ (id)sandboxBox;

/* instance methods */
- (_Bool)isValid;
- (id)dictionary;
- (id)valueForKey:(id)key;
- (id)urlForKey:(id)key;
- (id)initWithStoreClient:(id)client;
- (id)initWithBagType:(unsigned long long)type;
- (id)loadBagAndObserveUpdatesWithHandler:(id /* block */)handler;
- (_Bool)regexWithKey:(id)key matchesString:(id)string;
- (_Bool)sendGUIDWithURL:(id)url;
- (void)startUpdate;
- (void)stopObservingBagWithObserver:(id)observer;
- (id)storefrontURL;
- (_Bool)urlIsTrusted:(id)trusted;

@end

#endif /* CKBag_h */
```

**Reply by blacktop** [reply]
macOS 15.1

**Reply by rgoldberg** [reply]
@blacktop Thanks.

Can you share that command line that worked?

I'm not sure if you ran `ipsw dyld macho …`, `ipsw class-dump …`, or some other command. Nor do I know exactly what combo of arguments worked.

I'll let you know if that works on my Mac.

**Reply by rgoldberg** [reply]
@blacktop Thanks.

Can you share the command line that worked?

I'm not sure if you ran `ipsw dyld macho …`, `ipsw class-dump …`, or some other command. Nor do I know exactly what combo of arguments worked.

I'll let you know if that works on my Mac.

Thanks again.

**Reply by blacktop** [reply]
apologies: I ran
```bash
ipsw class-dump '/System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_arm64e' 'CommerceKit' --headers --output /tmp/CommerceKit_HDRS
```

**Reply by rgoldberg** [reply, marked as answer]
No problem. Thanks for all the help.

I ran the same command, except I had to change the DSC path, as the path you used did not exist on my file system. I assume that's due to macOS version differences.

```
ipsw class-dump /System/Library/dyld/dyld_shared_cache_arm64e 'CommerceKit' --headers --output /tmp/CommerceKit_HDRS
```

Contents of `CKBag.h` still have start-truncated symbol names, so I guess that I need a newer macOS.

```objective-c
//
//   Generated by https://github.com/blacktop/ipsw (Version: 3.1.544, BuildCommit: Homebrew)
//
//    - LC_BUILD_VERSION:  Platform: macOS, MinOS: 12.7, SDK: 12.7, Tools: [ld (756.90), ld (760)]
//    - LC_SOURCE_VERSION: 712.6.1.0.0
//
#ifndef CKBag_h
#define CKBag_h
@import Foundation;

#include "CKStoreClient.h"

@interface CKBag : NSObject

@property (readonly, nonatomic) CKStoreClient *storeClient;

/* class methods */
+ (id)itiveIntegerBuilder;
+ (id)ateLearningAnalytics17GroundTruthRecord;
+ (id)entifier;
+ (id)Array",R,C,N,V_strokeIdentifiers;
+ (id)s;

/* instance methods */
- (_Bool)CKDPRecordRetrieveChangesResponseRecordChange;
- (void)nURLRequestWrapperOperation;
- (id)rsonaPersistenceLayoutManager;
- (id)itiveIntegerBuilder;
- (id)ateLearningAnalytics17GroundTruthRecord;
- (id)BetweenRefreshes;
- (id)er",&,V_authenticationPluginManager;
- (id)_OSLogEventStoreTimeRef",R,N,V_oldestSignpost;
- (_Bool)rokeGroupType;
- (_Bool)T@"NSLocale",R,C,N,V_preferredLocale;
- (_Bool)edLocale;
- (id)ovider;
- (id)dentifierFromData:(id /* block */)data;
- (void)r:(id)r;
- (void)trokeIdentifier;

@end

#endif /* CKBag_h */
```

Thanks again for the help.

**Reply by blacktop** [reply]
`ipsw` focuses on supporting the latest versions of iOS/iPadOS/watchOS/macOS etc there are older versions that are NOT supported.  I believe when I started working on `ipsw` iOS 13 or iOS 14 was current? That is as far back as I really want to support, but if adding support isn't a lot of work I could add it.  I don't have time to explore macOS 12 right now, but if you want to make an GH issue so it is at least tracked you can and I'll try to get around to it at some point.

**Reply by rgoldberg** [reply]
Thanks for the offer, but it's probably not worth your time. Most people will be off of 12, and I can get someone else to generate the headers for me from a newer macOS.

**Reply by rgoldberg** [reply]
@blacktop I'm working on `mas`, which currently supports back to macOS 10.11. Parts of the private frameworks it uses have broken at various different macOS versions. Extract headers on older macOS versions would allow us to see which versions support which private functions, properties, etc., so we could make more parts of `mas` function on more macOS versions.

I just recently asked (in #575) if anyone has the extracted private framework headers for various macOD versions available online.

If such a header archive isn't available, it actually would make sense to try to see if you can make the command that didn't work on macOS 12 work on it & on other older macOS. If no one replies about such an archive after a little bit, I'll open an issue about making `ipsw` work on older macOS.

Thanks.
